cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -154,13 +155,48 @@ public void onSuccess(NodeData result) {
 
       @Override
       public void onFailure(Throwable t) {
-        routeConfigMap.remove(serviceId);
         settableFuture.setException(t);
+        // If no node is present, then simply place a watch on the node to monitor further data updates
+        // If its an exception other than NoNodeException, then remove the future, so that zkClient#getData
+        // will be retried during the next fetch config request
+        if (t instanceof KeeperException.NoNodeException) {
+          existsAndWatch(serviceId);
+        } else {
+          routeConfigMap.remove(serviceId);
```

Comment:
@chtyim In the case where fetching of data failed with an exception, if the failure was because there was no node present, then we place a watch on it. Since we set the exception, the future#get will return with that exception and we will return null as the RouteConfig. But in case the fetch of the data failed with any exception other than NoNodeException, we want to retry fetching data again and so remove the key (serviceId) from the cache - but we still set the exception and so the future#get will return. Is there anything else do you think we should do?